### PR TITLE
Refactor AGInfo and Imperative

### PR DESCRIPTION
**Pull request by larroy**
## Description ##

This refactor improves the code in the backward pass. Breaks a long function into smaller functions with more semantic variables that indicate what's happening and the intention. This makes the code hopefully bit easier to maintain.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by anirudhacharya** [comment]
@mxnet-label-bot add [pr-work-in-progress]

**Comment by anirudhacharya** [comment]
@mxnet-label-bot add [pr-work-in-progress]

**Comment by larroy** [comment]
@mxnet-label-bot add [Backend]

**Comment by szha** [comment]
@larroy could you share the purpose of this refactor?

File: include/mxnet/imperative.h [hunk 1]
```diff
@@ -35,48 +35,53 @@
...
-  /*! \brief */
-  class AGInfo {
-   public:
-    Context ctx;
-    OpReqType grad_req;
-    OpStatePtr state;
-    std::vector<NDArray> outputs;
-    std::vector<NDArray> out_grads;
-    bool fresh_out_grad;
+  Context ctx;
```
Comment (szha) [inline include/mxnet/imperative.h]: two spaces for indentation.

Comment (larroy) [reply include/mxnet/imperative.h]: Thanks, will fix.

**Comment by larroy** [comment]
@szha While working in an internal project which uses second order gradient we had very difficult to debug crashes which motivated to refactor that code which can be made easier to maintain and would benefit the project.

File: include/mxnet/imperative.h (hunk 1 above)
Comment (larroy) [reply include/mxnet/imperative.h]: It's indendeted at two spaces already, public at one, as in other files, is this not correct?

**Comment by Roshrini** [comment]
@larroy Can you please fix CI failure?

**Comment by larroy** [comment]
@szha should I fix this PR to get it merged or abandon it?

**Comment by szha** [comment]
@larroy let's work towards merging this. @eric-haibin-lin @junrushao1994 please take a look.

**Comment by larroy** [comment]
CI is passing smoothly.

**Comment by larroy** [comment]
@mxnet-label-bot update [Backend, pr-awaiting-review]

**Comment by larroy** [comment]
@mxnet-label-bot update [Backend, pr-awaiting-review]

**Comment by larroy** [comment]
Is this good to go?

**Comment by anirudhacharya** [comment]
@larroy why does this PR include changes to the tvm submodule?

**Comment by larroy** [comment]
@anirudhacharya it updates tvm subrepo.

File: src/imperative/imperative.cc [hunk 2]
```diff
@@ -124,29 +124,34 @@ void Imperative::MarkVariables(
...
+      grad_info.outputs.emplace_back(gradients[i]->Detach());
+      grad_info.ctx = gradients[i]->ctx();
+    }
   }
 }
 
 
+
+
+
```
[the remark this answers is not in the document]
Comment (larroy) [reply src/imperative/imperative.cc]: As we discussed f2f, we can merge it as it is, we can do these small improvements in subsequent PRS, either by you or by me.

**Comment by eric-haibin-lin** [comment]
looks good

**Comment by larroy** [comment]
Anything else needed for merge?

**Comment by larroy** [comment]
@mxnet-label-bot remove [pr-awaiting-review] add [pr-awaiting-merge]

**Comment by larroy** [comment]
@mxnet-label-bot add [pr-awaiting-merge]

File: src/imperative/imperative.cc [hunk 3]
```diff
@@ -124,29 +124,34 @@ void Imperative::MarkVariables(
...
-    info.out_grads.emplace_back(gradients[i]->Detach());
-    info.grad_req = static_cast<OpReqType>(grad_reqs[i]);
-    info.ctx = variables[i]->ctx();
-
-    gradients[i]->entry_ = nnvm::NodeEntry{
-        nnvm::Symbol::CreateVariable("grad" + str_c).outputs[0].node, 0, 0};
-    AGInfo& grad_info = AGInfo::Create(gradients[i]->entry_.node);
-    grad_info.outputs.emplace_back(gradients[i]->Detach());
-    grad_info.ctx = gradients[i]->ctx();
+    {
```
Comment (apeforest) [inline src/imperative/imperative.cc]: Why add a pair of braces here?

File: src/imperative/imperative.cc [hunk 4]
```diff
@@ -124,29 +124,34 @@ void Imperative::MarkVariables(
...
+      variables[i]->entry_ = nnvm::NodeEntry{
+          nnvm::Symbol::CreateVariable("var" + str_c).outputs[0].node, 0, 0};
+      AGInfo &info = AGInfo::Create(variables[i]->entry_.node);
+      info.outputs.emplace_back(variables[i]->Detach());
+      info.out_grads.emplace_back(gradients[i]->Detach());
+      info.grad_req = static_cast<OpReqType>(grad_reqs[i]);
+      info.ctx = variables[i]->ctx();
+    }
+    {
```
Comment (apeforest) [inline src/imperative/imperative.cc]: Why add a pair of braces here?

File: src/imperative/imperative.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (apeforest) [reply src/imperative/imperative.cc]: Why add these lines in this PR?

File: src/imperative/imperative.cc [hunk 5]
```diff
@@ -316,181 +312,223 @@ std::vector<NDArray*> Imperative::Backward(
...
+        info.fresh_out_grad = true;
+      }
     }
-    CHECK_GT(xs.size(), 0)
+    CHECK_GT(var_nodes.variable_nodes.size(), 0)
         << "There are no inputs in computation graph that require gradients.";
   }
+  return var_nodes;
+}
+
```
Comment (apeforest) [inline src/imperative/imperative.cc]: remove extra line

File: src/imperative/imperative.cc [hunk 6]
```diff
@@ -316,181 +312,223 @@ std::vector<NDArray*> Imperative::Backward(
...
+    bool create_graph) {
+  using namespace nnvm;
+  using namespace imperative;
+  static const std::vector<const Op*> zero_ops{Op::Get("zeros_like"), Op::Get("_zeros")};
+  static const Op* copy_op = Op::Get("_copy");
+
+  Graph graph;
+  graph.outputs = CreateForwardGraph(outputs);
+
+
```
Comment (apeforest) [inline src/imperative/imperative.cc]: remove extra line

**Comment by marcoabreu** [comment]
Is the submodule update required for this refactor? If not, I'd prefer if we don't update it as part of this PR.

**Comment by marcoabreu** [comment]
Not ready for merge. Still open comments from @apeforest and @anirudhacharya

File: src/imperative/imperative.cc [hunk 7]
```diff
@@ -232,16 +237,13 @@ void Imperative::RecordOp(
     node->inputs.resize(inputs.size());
   }
 
-  std::vector<bool>& save_inputs = *p_save_inputs;
```
Comment (apeforest) [inline src/imperative/imperative.cc]: I feel using these reference is better for readability. Also it's consistent with line https://github.com/apache/incubator-mxnet/pull/14836/files#diff-2d0bd6acfa276757ae59106dbed8e3e5R158

File: src/imperative/imperative.cc [hunk 8]
```diff
@@ -273,40 +275,34 @@ void Imperative::RecordOp(
...
-  using namespace nnvm;
-  using namespace imperative;
-  static const std::vector<const Op*> zero_ops{Op::Get("zeros_like"), Op::Get("_zeros")};
-  static const Op* copy_op = Op::Get("_copy");
-
-  // Construct forward graph
-  Graph graph;
-  graph.outputs.reserve(outputs.size());
-  for (const auto& i : outputs) {
+std::vector<nnvm::NodeEntry> Imperative::CreateForwardGraph(const std::vector<NDArray*>& outputs) {
```
Comment (apeforest) [inline src/imperative/imperative.cc]: I found this function and naming quite confusing. It is not really creating a graph, but simply assign a vector of outputs to the graph outputs. Maybe consider creating and returning the graph itself from this function?

File: src/imperative/imperative.cc [hunk 9]
```diff
@@ -316,181 +312,223 @@ std::vector<NDArray*> Imperative::Backward(
...
-  Symbol sym;
-  sym.outputs = graph.outputs;
-  std::vector<NodeEntry> xs;
-  std::vector<NDArray*> x_grads;
-  std::vector<OpReqType> x_reqs;
-  if (variables.size()) {
-    xs.reserve(variables.size());
-    x_grads.reserve(variables.size());
-    x_reqs.reserve(variables.size());
+struct Imperative::GradientVariableNodes {
```
Comment (apeforest) [inline src/imperative/imperative.cc]: Since these vector will always have the same size, would it make sense to create a struct like the following:
```suggestion
struct Imperative::GradientVariableNode {
     nnvm::NodeEntry variable_node;
     NDArray *gradient
     OpReqType op_req_type 
```
and return a vector of this structure from `CreateGradientVariableNodes` function?

File: src/imperative/imperative.cc [hunk 10]
```diff
@@ -316,181 +312,223 @@ std::vector<NDArray*> Imperative::Backward(
...
 
-  Graph g_graph = pass::MXGradient(
-      graph, graph.outputs, xs, ograd_entries,
+std::vector<NDArray*> Imperative::Backward(
+    const std::vector<NDArray*>& outputs,
+    const std::vector<NDArray*>& ograds,
+    const std::vector<NDArray*>& variables,
+    bool is_train, bool retain_graph,
+    bool create_graph) {
+  using namespace nnvm;
```
Comment (apeforest) [inline src/imperative/imperative.cc]: This namespace is redundant since in this function nnvm:: is still used

File: src/imperative/imperative.cc [hunk 11]
```diff
@@ -316,181 +312,223 @@ std::vector<NDArray*> Imperative::Backward(
...
+  using namespace imperative;
```
Comment (apeforest) [inline src/imperative/imperative.cc]: do we need this namespace?

File: src/imperative/imperative.cc [hunk 12]
```diff
@@ -316,181 +312,223 @@ std::vector<NDArray*> Imperative::Backward(
...
-  }
+
   if (create_graph) {
     states.resize(num_forward_nodes);
-    nnvm::DFSVisit(sym.outputs, [&](const nnvm::NodePtr& n) {
+    nnvm::DFSVisit(forward_outputs, [&](const nnvm::NodePtr& n) {
       AGInfo& info = AGInfo::Get(n);
-      states[idx.node_id(n.get())] = info.state;
+      states.at(indexed_graph.node_id(n.get())) = info.state;
       for (uint32_t i = 0; i < info.outputs.size(); ++i) {
```
Comment (apeforest) [inline src/imperative/imperative.cc]: change i to `size_t` to be consistent in your refactoring?

File: src/io/iter_libsvm.cc [hunk 13]
```diff
@@ -144,16 +144,16 @@ class LibSVMIter: public SparseIIterator<DataInst> {
     return true;
   }
 
-  virtual const DataInst &Value(void) const {
+  virtual const DataInst& Value(void) const {
     return out_;
   }
 
-  virtual const NDArrayStorageType GetStorageType(bool is_data) const {
+  virtual NDArrayStorageType GetStorageType(bool is_data) const {
```
Comment (apeforest) [inline src/io/iter_libsvm.cc]: Why remove const?

File: src/io/iter_libsvm.cc [hunk 14]
```diff
@@ -144,16 +144,16 @@ class LibSVMIter: public SparseIIterator<DataInst> {
...
     if (is_data) return kCSRStorage;
     return param_.label_shape.Size() > 1 ? kCSRStorage : kDefaultStorage;
   }
 
-  virtual const mxnet::TShape GetShape(bool is_data) const {
+  virtual mxnet::TShape GetShape(bool is_data) const {
```
Comment (apeforest) [inline src/io/iter_libsvm.cc]: Why remove const?

File: src/io/iter_prefetcher.h [hunk 15]
```diff
@@ -131,7 +131,7 @@ class PrefetcherIter : public IIterator<DataBatch> {
     }
     return iter.Next(&out_);
   }
-  virtual const DataBatch &Value(void) const {
+  virtual const DataBatch& Value(void) const {
```
Comment (apeforest) [inline src/io/iter_prefetcher.h]: I think both styles are fine: https://google.github.io/styleguide/cppguide.html#Pointer_and_Reference_Expressions

No need to change this

File: src/io/iter_sparse_batchloader.h [hunk 16]
```diff
@@ -100,15 +100,15 @@ class SparseBatchLoader : public BatchLoader, public SparseIIterator<TBlobBatch>
     return false;
   }
 
-  virtual const TBlobBatch &Value(void) const {
+  virtual const TBlobBatch& Value(void) const {
```
Comment (apeforest) [inline src/io/iter_sparse_batchloader.h]: I think both styles are fine: https://google.github.io/styleguide/cppguide.html#Pointer_and_Reference_Expressions

No need to change this

File: src/io/iter_sparse_prefetcher.h [hunk 17]
```diff
@@ -126,15 +126,15 @@ class SparsePrefetcherIter : public PrefetcherIter {
   virtual bool Next(void) {
     return PrefetcherIter::Next();
   }
-  virtual const DataBatch &Value(void) const {
```
Comment (apeforest) [inline src/io/iter_sparse_prefetcher.h]: I think both styles are fine: https://google.github.io/styleguide/cppguide.html#Pointer_and_Reference_Expressions

No need to change this

File: src/io/iter_sparse_prefetcher.h [hunk 18]
```diff
@@ -126,15 +126,15 @@ class SparsePrefetcherIter : public PrefetcherIter {
...
+  virtual const DataBatch& Value(void) const {
     return PrefetcherIter::Value();
   }
 
-  virtual const NDArrayStorageType GetStorageType(bool is_data) const {
+  virtual NDArrayStorageType GetStorageType(bool is_data) const {
```
Comment (apeforest) [inline src/io/iter_sparse_prefetcher.h]: Why remove const?

File: src/io/iter_sparse_prefetcher.h [hunk 19]
```diff
@@ -126,15 +126,15 @@ class SparsePrefetcherIter : public PrefetcherIter {
...
     return sparse_loader_->GetStorageType(is_data);
   }
 
-  virtual const mxnet::TShape GetShape(bool is_data) const {
+  virtual mxnet::TShape GetShape(bool is_data) const {
```
Comment (apeforest) [inline src/io/iter_sparse_prefetcher.h]: Why remove const?

File: src/nnvm/gradient.cc [hunk 20]
```diff
@@ -190,7 +190,8 @@ Graph Gradient(Graph src) {
       if (grad_fun_map.contains(ptr->op())) {
         input_grads = grad_fun_map[ptr->op()](fwd_node, out_agg_grads);
         CHECK_EQ((*rit)->inputs.size(), input_grads.size())
-            << "Gradient function not returning enough gradient";
+            << "Gradient function not returning enough gradient, there should be as many gradients"
```
Comment (apeforest) [inline src/nnvm/gradient.cc]: Nice comment!

File: tests/cpp/misc/libinfo_test.cc [hunk 21]
```diff
@@ -30,4 +30,5 @@ using namespace mxnet::features;
 TEST(RuntimeTest, RuntimeTestAll) {
   EXPECT_EQ(EnumNames::names.size(), MAX_FEATURES);
   const auto& features = LibInfo::getInstance()->getFeatures();
+  EXPECT_FALSE(features.empty());
```
Comment (apeforest) [inline tests/cpp/misc/libinfo_test.cc]: Is this change related to this refactoring?

File: src/executor/exec_pass.h [hunk 22]
```diff
@@ -258,6 +258,7 @@ inline Graph MXGradient(
   if (copy_op_str != std::string()) {
       graph.attrs["copy_op"] = std::make_shared<any>(std::move(copy_op_str));
   }
+  /// @sa nnvm::pass::Gradient in gradient.cc
```
Comment (apeforest) [inline src/executor/exec_pass.h]: What does @sa mean?

File: include/mxnet/io.h [hunk 23]
```diff
@@ -52,7 +52,7 @@ class IIterator : public dmlc::DataIter<DType> {
   /*! \brief move to next item */
   virtual bool Next(void) = 0;
   /*! \brief get current data */
-  virtual const DType &Value(void) const = 0;
```
Comment (apeforest) [inline include/mxnet/io.h]: I think both styles are fine: https://google.github.io/styleguide/cppguide.html#Pointer_and_Reference_Expressions

No need to change this

File: include/mxnet/imperative.h [hunk 24]
```diff
@@ -160,15 +165,37 @@ class Imperative {
 
  private:
   friend class NDArray;
-  /*! \brief make constructor protected. */
+  /*! Create a forward graph
+   * @param output_nodes graph node vector to add nodes to
+   * @param outputs source ndarrays
+   * @return vector of nodes
+   */
+  static std::vector<nnvm::NodeEntry> CreateForwardGraph(const std::vector<NDArray*>& outputs);
```
Comment (apeforest) [inline include/mxnet/imperative.h]: This function is very confusing. It does not really create a graph

File: src/executor/exec_pass.h (hunk 22 above)
Comment (larroy) [reply src/executor/exec_pass.h]: "see also" http://www.doxygen.nl/manual/commands.html  search "\sa"

File: tests/cpp/misc/libinfo_test.cc (hunk 21 above)
Comment (larroy) [reply tests/cpp/misc/libinfo_test.cc]: no, fixes a warning.

File: src/io/iter_sparse_prefetcher.h (hunk 19 above)
Comment (larroy) [reply src/io/iter_sparse_prefetcher.h]: What does const in a return value not by reference do? It's meaningless.

File: src/io/iter_sparse_prefetcher.h (hunk 18 above)
Comment (larroy) [reply src/io/iter_sparse_prefetcher.h]: I replied below. It makes very little sense.

File: src/io/iter_libsvm.cc (hunk 14 above)
Comment (larroy) [reply src/io/iter_libsvm.cc]: I replied below.

File: src/io/iter_libsvm.cc (hunk 13 above)
Comment (larroy) [reply src/io/iter_libsvm.cc]: I replied below.

File: src/imperative/imperative.cc (hunk 7 above)
Comment (larroy) [reply src/imperative/imperative.cc]: I don't think is consistent with requiring the use of pointers for output arguments, if you then change it to a reference there's no point on passing pointers in the first place. Please take that into consideration.

**Comment by marcoabreu** [comment]
Could you link the corresponding PR on the submodule  please?

File: src/imperative/imperative.cc (hunk 7 above)
Comment (larroy) [reply src/imperative/imperative.cc]: Please agree with @anirudhacharya on what you guys think is best. I think it comes down to "developer preference" as Tianqi would say. Your comments request in different directions.

File: src/imperative/imperative.cc (hunk 8 above)
Comment (larroy) [reply src/imperative/imperative.cc]: Done

File: src/imperative/imperative.cc (hunk 9 above)
Comment (larroy) [reply src/imperative/imperative.cc]: This is the old SOA vs AOS dilema, in general is less efficient to do AOS.
https://www.youtube.com/watch?v=Dj3lkfIEUyA
This is well known in gaming circles, as you loose quite a bit of cache efficiency due to AoS.

I don't see a real benefit of having a single array here.

File: src/imperative/imperative.cc (hunk 10 above)
Comment (larroy) [reply src/imperative/imperative.cc]: It's needed for NodeEntry and Graph to reduce noise, otherwise I wouldn't have typed using.

File: src/imperative/imperative.cc (hunk 11 above)
Comment (larroy) [reply src/imperative/imperative.cc]: yes

File: src/imperative/imperative.cc (hunk 12 above)
Comment (larroy) [reply src/imperative/imperative.cc]: Ok, but it had a reason for that.

File: src/io/iter_prefetcher.h (hunk 15 above)
Comment (larroy) [reply src/io/iter_prefetcher.h]: I find the former difficult to read, but ok.

**Comment by larroy** [comment]
https://github.com/dmlc/tvm/pull/3017

**Comment by larroy** [comment]
https://github.com/apache/incubator-mxnet/pull/14095

**Comment by larroy** [comment]
https://github.com/dmlc/tvm/pull/2576

File: src/imperative/imperative.cc (hunk 4 above)
Comment (larroy) [reply src/imperative/imperative.cc]: To separate the two clear parts in scopes. No gusta?

**Comment by marcoabreu** [comment]
Okay that looks reasonable, thanks :)

File: src/imperative/imperative.cc (hunk 10 above)
Comment (apeforest) [reply src/imperative/imperative.cc]: Then you may want to update this line: https://github.com/apache/incubator-mxnet/pull/14836/files#diff-2d0bd6acfa276757ae59106dbed8e3e5R384

**Comment by larroy** [comment]
@marcoabreu @szha @apeforest @anirudhacharya can you guys approve if you don't have any more comments?  Thanks.

**Comment by larroy** [comment]
@mxnet-label-bot add [pr-awaiting-review]

**Comment by larroy** [comment]
@mxnet-label-bot remove [pr-work-in-progress]

**Comment by larroy** [comment]
@mxnet-label-bot remove [pr-awaiting-response]

**Comment by eric-haibin-lin** [comment]
LGTM

File: src/imperative/imperative.cc (hunk 4 above)
Comment (apeforest) [reply src/imperative/imperative.cc]: No, it's better to create a function if you want to keep these variables local. This is not a recommended coding style

File: src/imperative/imperative.cc (hunk 9 above)
Comment (apeforest) [reply src/imperative/imperative.cc]: The benefit of keeping an array of struct is to make sure they have the same length. In your design, there is no direct way to guarantee this invariance.

File: src/imperative/imperative.cc (hunk 10 above)
Comment (apeforest) [reply src/imperative/imperative.cc]: If already using nnvm, then why nnvm::NodeEntry?

File: src/imperative/imperative.cc [hunk 25]
```diff
@@ -316,181 +308,220 @@ std::vector<NDArray*> Imperative::Backward(
...
+    arrays[eid] = gvars.gradients[i - num_forward_outputs];
     ref_count[eid] = 1;
   }
 
   // Assign context
-  auto vctx = PlaceDevice(idx);
+  auto vctx = PlaceDevice(indexed_graph);
 
   // Infer shape type
   {
```
Comment (apeforest) [inline src/imperative/imperative.cc]: Can we create a function to infershape here, instead of using braces.

Comment (apeforest) [reply src/imperative/imperative.cc]: Since this work itself is a refactoring, would it be better to perform a complete one?

File: src/imperative/imperative.cc (hunk 9 above)
Comment (apeforest) [reply src/imperative/imperative.cc]: The efficiency of AOS vs SOA does not apply here as you only perform adding element and no ALU related ops in the loop.

On the other hand, using a struct of vectors seems only move the original code to a struct, could we improve the readability during this refactoring?

**Comment by larroy** [comment]
I don't think the code review flow in this PR is productive or encouraging and I have spent already too much time on this. Closing.
